File: Baekjoon/1759.py
import sys
input = sys.stdin.readline

l, c = map(int, input().split())
mo = ['a','e','i','o','u']

arr = list(input().split())
arr.sort()
result = []

def dfs(idx, key, mo_num, ja_num):
  if idx >= c:
    return

  key += arr[idx] # 암호 추가
  status = "mo"
  if arr[idx] in mo: # 모음이면 모음 카운트 추가
    mo_num += 1
  else: # 자음이면 자음 카운트 추가
    ja_num += 1
    status = "ja"

  if len(key) == l and mo_num >= 1 and ja_num >= 2:
    result.append(key)

  dfs(idx+1, key, mo_num, ja_num) # 자기자신을 추가하는 경우
  if status == "mo":
    mo_num -= 1
  else:
    ja_num -= 1
  dfs(idx+1, key[:-1], mo_num, ja_num) # 자기자신을 추가하는 경우

dfs(0, "", 0, 0)

for i in range(len(result)):
  print(result[i])

File: Baekjoon/1759.java
import java.util.*;
import java.io.*;

public class Main {
    static int l,c;
    static Character[] arr;
    static Character[] mo = new Character[]{'a','e','i','o','u'};
    static List<String> result = new ArrayList<>();
    
    public static void main(String args[]) throws IOException{
      BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
      StringTokenizer st = new StringTokenizer(br.readLine());
      l = Integer.parseInt(st.nextToken());
      c = Integer.parseInt(st.nextToken());
      
      arr = new Character[c];
      
      st = new StringTokenizer(br.readLine());
      for(int i=0; i<c; i++){
          arr[i] = st.nextToken().charAt(0);
      }
      
      Arrays.sort(arr);
      
      dfs(0, "", 0, 0);
      
      for(String s : result){
          System.out.println(s);
      }
    }
    
    static void dfs(int idx, String key, int ja_num, int mo_num){
        if (idx >= c)
            return;
            
        key += (char) arr[idx];
        String chk = "mo";
        
        if (Arrays.asList(mo).contains(arr[idx])){
            mo_num++;
        } else{
            chk = "ja";
            ja_num++;
        }
        
        if(key.length() == l && mo_num >= 1 && ja_num >= 2){
            result.add(key);
        }
        
        dfs(idx + 1, key, ja_num, mo_num);
        if (chk.equals("mo")){
            dfs(idx + 1, key.substring(0,key.length() - 1), ja_num, mo_num - 1);
        } else{
            dfs(idx + 1, key.substring(0,key.length() - 1), ja_num - 1, mo_num);
        }
    }
}
